File: frontend/lib/models/space.dart
class Location {
  final String id;
  final String name;
  final bool isDefault;
  final int zoneCount;

  Location({required this.id, required this.name, this.isDefault = false, this.zoneCount = 0});

  factory Location.fromJson(Map<String, dynamic> json) => Location(
        id: json['id'],
        name: json['name'],
        isDefault: json['is_default'] ?? false,
        zoneCount: json['zone_count'] ?? 0,
      );
}

class Zone {
  final String id;
  final String locationId;
  final String name;
  final String? templateType;

  Zone({required this.id, required this.locationId, required this.name, this.templateType});

  factory Zone.fromJson(Map<String, dynamic> json) => Zone(
        id: json['id'],
        locationId: json['location_id'],
        name: json['name'],
        templateType: json['template_type'],
      );
}

class Container_ {
  final String id;
  final String zoneId;
  final String name;
  final List<Slot> slots;

  Container_({required this.id, required this.zoneId, required this.name, this.slots = const []});

  factory Container_.fromJson(Map<String, dynamic> json) => Container_(
        id: json['id'],
        zoneId: json['zone_id'],
        name: json['name'],
        slots: (json['slots'] as List?)?.map((s) => Slot.fromJson(s)).toList() ?? [],
      );
}

class Slot {
  final String id;
  final String containerId;
  final String name;
  final int level;
  final int itemCount;

  Slot({required this.id, required this.containerId, required this.name, this.level = 0, this.itemCount = 0});

  factory Slot.fromJson(Map<String, dynamic> json) => Slot(
        id: json['id'],
        containerId: json['container_id'],
        name: json['name'],
        level: json['level'] ?? 0,
        itemCount: json['item_count'] ?? 0,
      );
}

File: backend/app/schemas/space.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


# --- Location ---
class LocationCreate(BaseModel):
    name: str = Field(..., max_length=100)
    is_default: bool = False


class LocationResponse(BaseModel):
    id: str
    name: str
    is_default: bool
    created_at: datetime
    zone_count: int = 0

    model_config = {"from_attributes": True}


# --- Zone ---
class ZoneCreate(BaseModel):
    location_id: str
    name: str = Field(..., max_length=100)
    template_type: Optional[str] = None


class ZoneResponse(BaseModel):
    id: str
    location_id: str
    name: str
    template_type: Optional[str] = None

    model_config = {"from_attributes": True}


# --- Container ---
class SlotCreate(BaseModel):
    name: str = Field(..., max_length=100)
    level: int = 1


class ContainerCreate(BaseModel):
    zone_id: str
    name: str = Field(..., max_length=100)
    slots: list[SlotCreate] = []


class ContainerResponse(BaseModel):
    id: str
    zone_id: str
    name: str
    slots: list["SlotResponse"] = []

    model_config = {"from_attributes": True}


# --- Slot ---
class SlotResponse(BaseModel):
    id: str
    container_id: str
    name: str
    level: int
    item_count: int = 0

    model_config = {"from_attributes": True}


class SlotUpdate(BaseModel):
    name: Optional[str] = None
    level: Optional[int] = None


class SlotPathResponse(BaseModel):
    slot_id: str
    slot_name: str
    container_id: str
    container_name: str
    zone_id: str
    zone_name: str
    location_id: str
    location_name: str
    breadcrumb: str
